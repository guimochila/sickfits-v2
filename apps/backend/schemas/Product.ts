import { list } from '@keystone-6/core'
import { integer, relationship, select, text } from '@keystone-6/core/fields'
import { isSignedIn, rules } from '../access'

const Product = list({
  access: {
    operation: {
      create: isSignedIn,
    },
    filter: {
      query: rules.canReadProducts,
      update: rules.canManageProducts,
      delete: rules.canManageProducts,
    },
  },
  fields: {
    name: text({ validation: { isRequired: true } }),
    description: text({ ui: { displayMode: 'textarea' } }),
    photo: relationship({
      ref: 'ProductImage.product',
      ui: {
        displayMode: 'cards',
        cardFields: ['image', 'altText'],
        inlineCreate: { fields: ['image', 'altText'] },
        inlineEdit: { fields: ['image', 'altText'] },
      },
    }),
    status: select({
      options: [
        {
          label: 'Draft',
          value: 'DRAFT',
        },
        {
          label: 'Available',
          value: 'Available',
        },
        {
          label: 'Unavailable',
          value: 'UNAVAILABLE',
        },
      ],
      defaultValue: 'DRAFT',
      ui: {
        displayMode: 'segmented-control',
        createView: { fieldMode: 'hidden' },
      },
    }),
    price: integer(),
    user: relationship({
      ref: 'User.products',
      hooks: {
        resolveInput: ({ context }) => ({
          connect: { id: context.session.itemId },
        }),
      },
    }),
  },
})

export { Product }
