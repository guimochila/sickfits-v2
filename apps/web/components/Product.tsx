import Image from 'next/image'
import Link from 'next/link'
import type { Product as TProduct } from '../@types/graphql-generated'
import formatMoney from '../utils/formatMoney'
import Item from './styled/Item'
import PriceTag from './styled/PriceTag'
import Title from './styled/Title'

interface IProductProps {
  product: TProduct
}

function Product({ product }: IProductProps) {
  return (
    <Item>
      {product.name}
      <Image
        src={product.photo?.image?.publicUrlTransformed || ''}
        alt={product.name || 'Product image'}
        width={250}
        height={250}
      />
      <Title>
        <Link href={`/product/${product.id}`}>{product.name}</Link>
      </Title>
      <PriceTag>{formatMoney(product.price!)}</PriceTag>
      <p>{product.description}</p>
    </Item>
  )
}

export default Product
