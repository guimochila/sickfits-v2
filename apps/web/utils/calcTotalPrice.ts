import { CartItem } from '../@types/graphql-generated'

function calcTotalPrice(cart: CartItem[]) {
  return cart.reduce((total, cartItem) => {
    if (!cartItem.product) return total

    return total + cartItem.quantity * cartItem.product.price
  }, 0)
}

export default calcTotalPrice
