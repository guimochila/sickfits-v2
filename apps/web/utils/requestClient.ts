import { GraphQLClient } from 'graphql-request'
import { API_ENDPOINT } from '../config'

const gqlClient = new GraphQLClient(API_ENDPOINT, {
  credentials: 'include',
  mode: 'cors',
})

export default gqlClient
