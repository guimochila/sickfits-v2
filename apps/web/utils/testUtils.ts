import casual from 'casual'
import { PAGINATION_QUERY } from '../hooks/useProductsCount'
// seed it so we get consistent results
casual.seed(777)

const fakeItem = (): any => ({
  // __typename: 'Product',
  id: 'abc123',
  price: 5000,
  user: null,
  photo: {
    id: 'abc123',
    altText: 'dogs are best',
    image: {
      publicUrlTransformed: 'http://localhost/dog.jpg',
      id: '5e2a13ff689b2835ae71d1a7',
      filename: 'dog.jpg',
      originalFilename: 'dog.jpg',
      mimetype: 'image/jpeg',
      encoding: '7bit',
      publicUrl: 'http://localhost/dog.jpg',
    },
  },
  name: 'dogs are best',
  description: 'dogs',
  status: 'AVAILABLE',
})

const fakeUser = (overrides = {}) => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
  orders: [],
  cart: [],
  ...overrides,
})

const fakeOrderItem = () => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: {
    image: `${casual.word}.jpg`,
  },
  name: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
})

const fakeOrder = () => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2022-12-11T20:16:13.797Z',
  user: fakeUser(),
})

const fakeCartItem = (overrides = {}) => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  product: fakeItem(),
  user: fakeUser(),
  ...overrides,
})

// Fake LocalStorage
class LocalStorageMock {
  store: any

  constructor() {
    this.store = {}
  }

  clear() {
    this.store = {}
  }

  getItem(key: string) {
    return this.store[key] || null
  }

  setItem(key: string, value: any) {
    this.store[key] = value.toString()
  }

  removeItem(key: string) {
    delete this.store[key]
  }
}

function makePaginationMocksFor(length: number) {
  return [
    {
      request: { query: PAGINATION_QUERY },
      result: {
        data: {
          _allProductsMeta: {
            count: length,
          },
          itemsConnection: {
            __typename: 'aggregate',
            aggregate: {
              count: length,
              __typename: 'count',
            },
          },
        },
      },
    },
  ]
}

export {
  makePaginationMocksFor,
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
}
