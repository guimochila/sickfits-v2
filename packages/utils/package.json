{
  "name": "@sickfits/utils",
  "private": true,
  "version": "0.0.1",
  "main": "build/index.js",
  "types": "build/index.d.ts",
  "license": "MIT",
  "author": {
    "name": "Guilherme Scaldelai"
  },
  "devDependencies": {
    "@types/jest": "^28.1.0",
    "babel-jest": "^28.1.0",
    "jest": "^28.1.0",
    "ts-jest": "^28.0.4",
    "typescript": "^4.6.2"
  },
  "scripts": {
    "build": "tsc",
    "test": "NODE_ENV=test jest",
    "test:watch": "NODE_ENV=test jest --watch"
  }
}
