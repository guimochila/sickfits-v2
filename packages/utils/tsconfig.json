{
  "compilerOptions": {
    "target": "es5",
    "module": "commonjs",
    "declaration": true,
    "outDir": "./build",
    "strict": true,
    "skipLibCheck": true
  },
  "include": ["src/**/*"],
  "exclude": ["node_modules", "src/**/*.test.ts"]
}
