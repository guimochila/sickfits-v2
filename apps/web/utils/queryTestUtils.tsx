import { render } from '@testing-library/react'
import { graphql } from 'msw'
import * as React from 'react'
import { QueryClient, QueryClientProvider } from 'react-query'
import { fakeCartItem, fakeItem, fakeUser } from './testUtils'

const product = fakeItem()

export const handlers = [
  // Single Product Query mock
  graphql.query('SINGLE_PRODUCT_QUERY', (req, res, ctx) => {
    const { id } = req.variables

    if (id === product.id) {
      return res(
        ctx.data({
          product,
        }),
      )
    }

    return res(ctx.errors([{ message: 'Item not found!' }]))
  }),

  // Current User Query mock
  graphql.query('CURRENT_USER_QUERY', (req, res, ctx) => {
    const isAuthenticated = sessionStorage.getItem('is-authenticated')

    if (!isAuthenticated) {
      return res(
        ctx.data({
          authenticatedItem: fakeUser(),
        }),
      )
    }

    if (isAuthenticated === 'yesWithCartItems') {
      return res(
        ctx.data({
          authenticatedItem: fakeUser({
            cart: [fakeCartItem()],
          }),
        }),
      )
    }

    return res(
      ctx.data({
        authenticatedItem: fakeUser(),
      }),
    )
  }),

  // Pagination mock
  graphql.query('PAGINATION_QUERY', (req, res, ctx) => {
    return res(
      ctx.data({
        productsCount: 18,
      }),
    )
  }),

  // Signup Mutation mock
  graphql.mutation('SIGNUP_MUTATION', (req, res, ctx) => {
    const { name, email } = req.variables

    return res(
      ctx.data({
        createUser: {
          name,
          email,
        },
      }),
    )
  }),

  // Reset Request Mutation mock
  graphql.mutation('REQUEST_RESET_MUTATION', (req, res, ctx) => {
    const { email } = req.variables

    if (email) {
      return res(
        ctx.data({
          sendUserPasswordResetLink: true,
        }),
      )
    }

    return res(
      ctx.data({
        sendUserPasswordResetLink: null,
      }),
    )
  }),

  // Create Mutation mock
  graphql.mutation('CREATE_PRODUCT_MUTATION', (req, res, ctx) => {
    const { name, price, description } = req.variables

    if (name && price && description) {
      return res(
        ctx.data({
          createProduct: fakeItem(),
        }),
      )
    }

    return res(ctx.errors([{ message: 'Something went wrong' }]))
  }),
]

const createTestQueryClient = () =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })

export function renderWithClient(ui: React.ReactElement) {
  const testQueryClient = createTestQueryClient()
  const { rerender, ...result } = render(
    <QueryClientProvider client={testQueryClient}>{ui}</QueryClientProvider>,
  )
  return {
    ...result,
    rerender: (rerenderUi: React.ReactElement) =>
      rerender(
        <QueryClientProvider client={testQueryClient}>
          {rerenderUi}
        </QueryClientProvider>,
      ),
  }
}

export function createWrapper() {
  const testQueryClient = createTestQueryClient()
  // eslint-disable-next-line react/display-name
  return ({ children }: { children: React.ReactNode }) => (
    <QueryClientProvider client={testQueryClient}>
      {children}
    </QueryClientProvider>
  )
}
