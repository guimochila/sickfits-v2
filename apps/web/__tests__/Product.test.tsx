import Product from '../components/Product'
import { renderWithClient } from '../utils/queryTestUtils'
import { screen } from '@testing-library/react'
import { fakeItem } from '../utils/testUtils'

const product = fakeItem()

describe('<Product/>', () => {
  it('renders out the price tag and title', () => {
    const { container } = renderWithClient(<Product product={product} />)
    const priceTag = screen.getByText('â‚¬50')
    expect(priceTag).toBeInTheDocument()

    const link = container.querySelector('a')
    expect(link).toHaveAttribute('href', '/product/abc123')
    expect(link).toHaveTextContent(product.name)
  })

  it('Renders and matches the snapshot', () => {
    const { container } = renderWithClient(<Product product={product} />)

    expect(container).toMatchSnapshot()
  })

  it('renders the image properly', () => {
    const { container } = renderWithClient(<Product product={product} />)

    const img = screen.getByAltText(product.name)
    expect(img).toBeInTheDocument()
  })
})
