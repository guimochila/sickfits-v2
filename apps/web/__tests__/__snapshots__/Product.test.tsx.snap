// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product/> Renders and matches the snapshot 1`] = `
<div>
  <div
    class="css-ig2a7j"
  >
    dogs are best
    <span
      style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
    >
      <span
        style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
      >
        <img
          alt=""
          aria-hidden="true"
          src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27250%27%20height=%27250%27/%3e"
          style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
        />
      </span>
      <img
        alt="dogs are best"
        data-nimg="intrinsic"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
      />
      <noscript />
    </span>
    <h3
      class="css-1ji38yg"
    >
      <a
        href="/product/abc123"
      >
        dogs are best
      </a>
    </h3>
    <span
      class="css-8u9v7m"
    >
      ‚Ç¨50
    </span>
    <p>
      dogs
    </p>
    <div
      class="buttonList"
    >
      <a
        href="/update?id=abc123"
      >
        Edit ‚úèÔ∏è
      </a>
      <button
        type="button"
      >
        Add
         To Cart üõí
      </button>
      <button
        type="button"
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;
